### PR TITLE
Refactor Companion Update Flow for robustness

### LINKED ISSUE #252: Companion Update Flow might fail silently
This issue is related to the scenario in https://github.com/paritytech/polkadot/pull/2610

Errors were found upon further inspection of the Google Cloud logs, therefore weakening the argument for https://github.com/paritytech/polkadot/pull/2610#issuecomment-800252606 being in the right trail *for that situation* (it'd be part of another problem: #236).

This issue therefore lists findings about the current problems with the Companion Update Flow not related to how the build system currently manages concurrent merge requests, but instead centered around how failure is badly handled currently. While #236 and #250 can't be solved reliably in the short term, what we can do for the time being is post a comment when an error happens so that developers can resolve them manually.

# Problems

## Cloning might fail

In https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/companion.rs#L65 a `git clone` is attempted. It might fail, and it indeed failed in that situation, although the error didn't matter. Either way, failure is not handled currently, but it should be. From the logs:

```
fatal: destination path 'polkadot' already exists and is not an empty directory
```

## Merge might be done on a outdated master branch

The merge master command does not fetch `origin/master` before attempting the merge, although it should, since it's only guaranteed to be up-to-date on a fresh clone (which might fail, as described above).

https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/companion.rs#L120

It showed up in the logs:

```
{
  "textPayload": "[2021-03-16T12:04:05Z INFO  parity_processbot::companion] Merging master.\n",
},
{
  "textPayload": "Already up to date.\n",
}
```

Regardless of whether or not `origin/master` was indeed up-to-date with Github at that point in time, this measure still needs to be taken into consideration.

## The bot fails silently

The logs shows that the [cargo update command](https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/companion.rs#L133) had failed:

```
{
  "textPayload": "Updating git repository `https://github.com/paritytech/substrate`\n",
},
{
  "textPayload": "Updating crates.io index\n",
},
{
  "textPayload": "error: no matching package named `sp-election-providers` found\n",
}
```

Due to the bugs mentioned before, no change was made to the repository, which meant there was nothing to commit

```
  "textPayload": "[2021-03-16T12:04:18Z INFO  parity_processbot::companion] Committing changes.\n",
},
{
  "textPayload": "On branch gav-abstract-payout-curve\n",
},
{
  "textPayload": "Your branch is up to date with 'temp/gav-abstract-payout-curve'.\n",
},
```

Even if those commands were properly checked for their exit status, as it currently is the bot would still fail silently with a single uninformative comment (as in https://github.com/paritytech/polkadot/pull/2610#issuecomment-800203132) because those errors are simply logged without posting a comment

https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/webhook.rs#L1281-L1284

# Solution

- Properly check all commands for errors
- Make the flow more robust (e.g. handle the case where the repository was already cloned without blindly ignoring the error)
- When the update flow fails, post a comment for context

### PR DESCRIPTION
**Pull request by joao-paulo-parity**
See #252 for context. Unfortunately we don't yet have an integration test suite to be confident about putting those changes in production. Regardless, I thought I should still have it ready here in case there's interest in trying it out with the staging environments, but otherwise this can come along with the PR for #251.

Motivation

- No need for manual logging before running commands as the `run_cmd` commands already do it
- The cleanup checks are now done eagerly instead of the old cleanup routine at the end of the update
- Meticulously check for errors and communicate them in the merge request

Tested scenarios

- https://github.com/paritytech/polkadot-for-processbot-staging/pull/5
- https://github.com/paritytech/polkadot-for-processbot-staging/pull/6

Google Cloud logs are more readable now:

```
Run git ["clone", "-v", "https://x-access-token:${SECRET}@github.com/paritytech/polkadot-for-processbot-staging.git"] in the current directory

Run git ["remote", "get-url", "paritytech"] in ./polkadot-for-processbot-staging

Run git ["remote", "add", "paritytech", "https://x-access-token:${SECRET}@github.com/paritytech/polkadot-for-processbot-staging.git"] in ./polkadot-for-processbot-staging

Run git ["fetch", "paritytech", "test_merge_commit"] in ./polkadot-for-processbot-staging

Run git ["branch", "-D", "test_merge_commit"] in ./polkadot-for-processbot-staging

Run git ["checkout", "--track", "paritytech/test_merge_commit"] in ./polkadot-for-processbot-staging

Run git ["fetch", "origin", "test_merge_commit_owner"] in ./polkadot-for-processbot-staging

Run git ["merge", "origin/test_merge_commit_owner", "--no-ff", "--no-edit"] in ./polkadot-for-processbot-staging

Run cargo ["update", "-vp", "sp-io"] in ./polkadot-for-processbot-staging

Run git ["status", "--short"] in ./polkadot-for-processbot-staging

Run git ["commit", "-am", "update Substrate"] in ./polkadot-for-processbot-staging

Run git ["push", "paritytech", "test_merge_commit"] in ./polkadot-for-processbot-staging

Getting the head SHA after a companion update in paritytech/test_merge_commit

Run git ["rev-parse", "HEAD"] in ./polkadot-for-processbot-staging

Companion updated; waiting for checks on github.com/paritytech/polkadot-for-processbot-staging/pull/6
```

Closes #252

### CODE REVIEW

File: src/companion.rs
Comment (haikoschol) [inline src/companion.rs:77]: This could be any random user/org and repository, not necessarily inside `paritytech`, right? In fact, I can't think of a reason for someone to clone `paritytech/polkadot` to `paritytech/foobar` and then submit PRs from that to `paritytech/polkadot`.

I vaguely remember that this whole bot-driven companion workflow cannot be used for PRs from forks, because those won't be run in CI or something like that. Is that still the case? Or did I remember that wrong?

Looking at the code that is deleted in this PR, it seems to have created a remote called `temp` which also points to the "owner" repo (so e.g. `paritytech/polkadot`?). I don't understand why it was done that way. Whatever the reason, the new code creates a remote pointing at a 3rd party repo (so e.g. `bob/polkadot`), named after the user or organization where a PR originates. And further down (line 201) new commits are pushed to that origin. In what scenarios would that succeed or fail? The bot usually doesn't have push permissions to `bob/polkadot`, right?

Comment (joao-paulo-parity) [reply src/companion.rs:77]: > This could be any random user/org and repository, not necessarily inside paritytech, right?

That's right. There's nothing blocking it from being started for any organization, although the expect scenario is a Substrate PR trying a companion update on Polkadot; in fact, it's hardcoded to only work for Substrate.

https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/webhook.rs#L1205

This code assumes that the bot will be installed in the companion PR's repository (`polkadot` in this case) since, as you noted, it needs push permissions to push commits to the companion.

---

> I vaguely remember that this whole bot-driven companion workflow cannot be used for PRs from forks, because those won't be run in CI or something like that. Is that still the case? Or did I remember that wrong?

I would expect the merge and push to work granted the fork's PR has [enabled edits from maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) since the bot should have the relevant permissions for being classified as a "maintainer".

For CI [the bot simply checks that everything is green](https://github.com/paritytech/parity-processbot/blob/dee4d2479cdbbf835160b26712203455b39eb21b/src/webhook.rs#L252) without regards from where those statuses came from, therefore I would expect it to work fine even if the branch does not belong to a `paritytech` repository as CI also runs on PRs from forks.

To answer the question more directly: this scenario is not taken into account currently, nor tested, so there's no certain response.

---

> Looking at the code that is deleted in this PR, it seems to have created a remote called temp which also points to the "owner"

`temp` was what's now called the contributor's MR. In the case of e.g. https://github.com/paritytech/polkadot/pull/2610 it's `paritytech/polkadot:gav-abstract-payout-curve` (you can hover over the name to see it).

image

Failure or success is related to the bot having push permissions **to the PR**, not necessarily the contributor's repository. In that case the branch was created directly in `paritytech/polkadot`, where the bot is assumed to be installed, so it'll always succeed in that regard. I expect it to also work for forks' PRs granted they have the enabled "edits from maintainers" as I mentioned above, although that's really not taken into account currently.

Effectively the refactored code should achieve the same as the deleted code; most noticeably the naming has been made more meaningful (from `temp` to `contributor`) and made easier to follow.

> I don't understand why it was done that way

The point of this flow is twofold

- Create a master merge commit
- Update `Cargo.lock` to the latest SHA on Github after the companion on Substrate is merged

The bot has to commit those changes on the PR's author behalf. `temp` was pointing at the PR author's (here called the "contributor") repository and the new code should still be doing the same; it *just happens* that the "owner" and "contributor" are both `paritytech` in the case of e.g. https://github.com/paritytech/polkadot/pull/2610 but there's nothing in the code stopping it from running it on forks.

**Review by haikoschol — APPROVED** [review]

**Pull request merged**
